### PR TITLE
modules/job-manager:  [cleanup] simplify queue listing and refactor internal context

### PR DESCRIPTION
**Pull request by garlick**
I unintentionally created a bit of a mess in the job-manager by trying to tightly control communication between subsystems.  That created a situation where each subsystem had to acquire and cache pointers to the "contexts" of other subsystems, leading to a lot of duplication.

This restructures the module so that there is  one main context shared by all subsystems.  While each subsystem still contributes an opaque type to the context containing its private members, the opaque type is at least available to other subsystems via the main context.

This PR also gets rid of the "queue" pseudo container, which felt like a misstep.  Track active jobs directly in a zhashx_t, and the queue of pending jobs in a zlistx_t.  Some helpers for these were added to job.c.

Finally, the `job-manager.list` RPC is simplified and converted to a test-only interface now that `job-info.list` has taken over that function for front end tools.

### CODE REVIEW

**Comment by garlick** [comment]
Restarted one builder that hit a valgrind timeout.

**Comment by garlick** [comment]
One thing I think I should do here is move the "job hash" and "job list" helpers (pretty minimal) to `src/common/libjob` so they can be shared internally by components that need to maintain hashes or lists of jobs, using their own job object.

**Comment by chu11** [comment]
> One thing I think I should do here is move the "job hash" and "job list" helpers (pretty minimal) to src/common/libjob so they can be shared internally by components that need to maintain hashes or lists of jobs, using their own job object.

Good idea!  I think `job-info` could use them.

File: src/modules/job-manager/submit.c [hunk 1]
```diff
@@ -51,14 +50,14 @@ int submit_enqueue_one_job (struct queue *queue, zlist_t *newjobs, json_t *o)
         job_decref (job);
         return -1;
     }
-    if (queue_insert (queue, job, &job->queue_handle) < 0) {
+    if (zhashx_insert (active_jobs, &job->id, job) < 0) {
         job_decref (job);
         // EEXIST is not an error - there is a window for restart_from_kvs()
```
Comment (chu11) [inline src/modules/job-manager/submit.c]: perhaps this comment needs to be updated given code changes around it (i.e. zhashx_insert doesn't set EEXIST).

**Comment by garlick** [comment]
Just forced a push that
* fixes comment per @chu11's review
* moves `job_hash_create()` to libjob for local (internal) use
* use that in `job-manager`, `job-info` and `job-exec`
* rebase on current master

I couldn't generalize the zlistx_t since sort order depends on access to internal struct members, and there's not much left after that.

File: src/modules/job-manager/submit.c [hunk 2]
```diff
@@ -53,14 +50,16 @@ int submit_enqueue_one_job (struct queue *queue, zlist_t *newjobs, json_t *o)
         job_decref (job);
         return -1;
     }
-    if (queue_insert (queue, job, &job->queue_handle) < 0) {
+    if (zhashx_insert (active_jobs, &job->id, job) < 0) {
         job_decref (job);
-        // EEXIST is not an error - there is a window for restart_from_kvs()
-        // to pick up a job that also has a submit request in flight.
-        return (errno == EEXIST ? 0 : -1);
+        /* xhashx_insert() fails if hash item already exists.
```
Comment (chu11) [inline src/modules/job-manager/submit.c]: you mean zhashx_insert()? :-)

**Comment by garlick** [comment]
Doh!  Forced a push.

**Comment by garlick** [comment]
Thanks!

**Pull request merged**